libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1006,6 +1008,152 @@ std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
+				if gt(newLen, <maxArrayLength>) {
+					<panic>()
+				}
+
+				let oldLen := <fetchLength>(array)
+
+				if gt(newLen, oldLen) {
+					<increaseSize>(array, oldLen, newLen)
+				}
+
+				if lt(newLen, oldLen) {
+					<decreaseSize>(array, oldLen, newLen)
+				}
+			})")
+			("functionName", functionName)
+			("panic", panicFunction())
+			("fetchLength", arrayLengthFunction(_type))
+			("maxArrayLength", (u256(1) << 64).str())
+			("decreaseSize", decreaseByteArraySizeFunction(_type))
+			("increaseSize", increaseByteArraySizeFunction(_type))
+			.render();
+	});
+}
+
+string YulUtilFunctions::decreaseByteArraySizeFunction(ArrayType const& _type)
+{
+	string functionName = "byte_array_decrease_size_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, oldLen, newLen) {
+				switch lt(newLen, 32)
+				case  0 {
+					let arrayDataStart := <dataPosition>(array)
+					let deleteStart := add(arrayDataStart, div(add(newLen, 31), 32))
+
+					// we have to partially clear last slot that is still used
+					let offset := mod(newLen, 32)
+					if gt(offset, 0) { <partialClearStorageSlot>(sub(deleteStart, 1), offset) }
+
+					<clearStorageRange>(deleteStart, add(arrayDataStart, div(add(oldLen, 31), 32)))
+
+					sstore(array, or(<shl>(1, newLen), 1))
+				}
+				default {
+					switch gt(oldLen, 31)
+					case 0 {
+						let arrayDataStart := <dataPosition>(array)
+						// clear whole old array, as we are transforming to short bytes array
+						<clearStorageRange>(arrayDataStart, add(arrayDataStart, div(add(oldLen, 31), 32)))
+						<transitLongToShort>(array, newLen)
+					}
+					default {
+						<changeLengthShort>(array, sload(array), newLen)
+					}
+				}
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("partialClearStorageSlot", partialClearStorageSlotFunction())
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			("shl", shiftLeftFunctionDynamic())
+			("transitLongToShort", byteArrayTransitLongToShortFunction(_type))
+			("changeLengthShort", changeLengthOfShortByteArrayFunction(_type))
+			.render();
+	});
+}
+
+string YulUtilFunctions::increaseByteArraySizeFunction(ArrayType const& _type)
+{
+	string functionName = "byte_array_increase_size_to_unpacked_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, oldLen, newLen) {
+				switch lt(oldLen, 32)
+				case 0 {
+					// in this case array stays unpacked, so we just set new length
+					sstore(array, or(<shl>(1, newLen), 1))
+				}
+				default {
+					let data := and(not(0xff), sload(array))
+					switch lt(newLen, 32)
+					case 0 {
+						// we need to copy elements to data area as we changed array from packed to unpacked
+						sstore(<dataPosition>(array), data)
+						sstore(array, or(<shl>(1, newLen), 1))
```

Comment:
`mul(2, newLen)`?

Reply:
I would say yes, I did `shl` as @ekpyron likes it more :-)

Reply:
Did I ask for it, or did you just anticipate me doing that :-D? I'm fine with ``mul(2, newLen)`` in any case.

Reply:
Although then I'd actually consider making the ``or`` an ``add`` as well... mixing bitwise and arithmetic is harder to read than either one, isn't it? And for the cost it shouldn't matter what we do here (I think without really having checked).

Reply:
The main reason is that especially the dynamic version is really expensive if we do not have a shl in the evm dialect.